### PR TITLE
feat: Install Poetry requires-plugins before running Poetry commands

### PR DESCRIPTION
**Pull request by markhallen**
### What are you trying to accomplish?

Poetry v2 supports a [`requires-plugins`](https://python-poetry.org/docs/pyproject/#requires-plugins) field in `pyproject.toml` that declares which Poetry plugins a project depends on. When Dependabot runs Poetry commands (lock, update, resolve) on a project that requires plugins, those plugins must be installed first — otherwise Poetry may fail with confusing errors or produce incorrect results.

This builds on the `requires-poetry` constraint support merged in #14684 by adding automatic plugin installation.

### Anything you want to highlight for special attention from reviewers?

- **New `PoetryPluginInstaller` class** — reads `[tool.poetry.requires-plugins]` from `pyproject.toml` and installs each declared plugin via `poetry self add`. Input validation (regex allowlists for plugin names and version constraints) prevents command injection. Uses `Shellwords.shellescape` as an additional safety layer.
- **Integrated into all three Poetry code paths** — the lockfile generator, file updater, and version resolver all call `poetry_plugin_installer.install_required_plugins` before executing Poetry commands, following the same pattern as `language_version_manager.install_required_python`.
- **Plugin installation failures are non-fatal** — if a plugin can't be installed (unavailable on PyPI, version conflict, private registry), a warning is logged and processing continues. The Poetry command itself may still succeed, or fail with a clearer error.
- **PEP 621 updater fallback** — `Pep621Updater` now falls back to matching against the normalized requirement string when `source_requirement` metadata is absent (e.g. after `DependencySet` merge or deserialization), preventing silent update failures.
- **Explicit `!=` operator handling** — `bump_single_requirement` now handles `!=` explicitly (keeps unchanged) instead of relying on a catch-all `else` branch.

### How will you know you've accomplished your goal?

- `PoetryPluginInstaller` has comprehensive unit tests covering: single plugin, multiple plugins, missing/empty section, invalid plugin names and constraints (command injection prevention), TOML parse errors, failure resilience (warns instead of raising), and idempotent installation.
- Integration tests in lockfile generator, file updater, and version resolver specs verify that `install_required_plugins` is invoked during their respective workflows.
- PEP 621 fallback test confirms dependencies without `source_requirement` metadata still get updated correctly.
- Range requirement edge case tests cover strict `>` lower bound conversion and `!=` exclusion alongside ranges for both `BumpVersions` and `BumpVersionsIfNecessary` strategies.
- 3 fixture files (`requires_plugins.toml`, `requires_plugins_multiple.toml`, `requires_plugins_none.toml`) support the test scenarios.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: python/spec/dependabot/python/update_checker/poetry_version_resolver_spec.rb [hunk 1]
```diff
@@ -665,4 +665,32 @@
...
+
+  describe "plugin installation during resolution" do
+    it "invokes the poetry plugin installer" do
+      plugin_installer = instance_double(
+        Dependabot::Python::PoetryPluginInstaller,
+        install_required_plugins: nil
+      )
+      allow(Dependabot::Python::PoetryPluginInstaller)
+        .to receive(:from_dependency_files).and_return(plugin_installer)
+      allow(Dependabot::SharedHelpers).to receive(:in_a_temporary_repo_directory).and_yield
```
[the remark this answers is not in the document]
Comment (markhallen) [reply python/spec/dependabot/python/update_checker/poetry_version_resolver_spec.rb]: Fixed in 12091ab. Updated to stub `SharedHelpers.in_a_temporary_directory` and `SharedHelpers.with_git_configured` instead of `SharedHelpers.in_a_temporary_repo_directory`, matching what `PoetryVersionResolver#fetch_latest_resolvable_version_string` actually calls.

File: python/spec/dependabot/python/file_updater/poetry_file_updater_spec.rb [hunk 2]
```diff
@@ -959,6 +959,40 @@
...
 
+  describe "plugin installation during update" do
+    it "invokes the poetry plugin installer" do
+      plugin_installer = instance_double(
+        Dependabot::Python::PoetryPluginInstaller,
+        install_required_plugins: nil
+      )
+      allow(Dependabot::Python::PoetryPluginInstaller)
+        .to receive(:from_dependency_files).and_return(plugin_installer)
+      allow(Dependabot::SharedHelpers).to receive(:in_a_temporary_repo_directory).and_yield
```
[the remark this answers is not in the document]
Comment (markhallen) [reply python/spec/dependabot/python/file_updater/poetry_file_updater_spec.rb]: Fixed in 12091ab. The tests were stubbing `SharedHelpers.in_a_temporary_repo_directory`, but the production code in `PoetryFileUpdater#updated_lockfile_content_for` uses `SharedHelpers.in_a_temporary_directory` + `SharedHelpers.with_git_configured`. Updated to stub both correct helpers, and also added `python_version` to the `LanguageVersionManager` double since `prepared_pyproject` calls `language_version_manager.python_version` before reaching the temporary directory block.

**Review by robaiken — DISMISSED** [review]

**Review by robaiken — DISMISSED** [review]

**Comment by markhallen** [comment]
### Post-deploy observation: `Expected lockfile to change!` error

After deploying this branch, we observed a new error in production:

```
RuntimeError: Expected lockfile to change!
```

**Root cause**: The `RequirementsUpdater` changes removed the `new_version_satisfies?` guard from `update_pyproject_version` for range requirements with the `BumpVersions` strategy. The new `bump_requirements_range` method now bumps lower bounds (e.g., `>=0.2.0,<1.0.0` → `>=0.3.5,<1.0.0`) even when the locked version already satisfies the original range.

This creates a scenario where:
1. The pyproject.toml is updated (constraint is bumped) → `file_changed?` returns true
2. `poetry update --lock` runs, but the dependency is already locked at the target version
3. The lockfile content is identical → `"Expected lockfile to change!"` is raised

**Fix** (commit `8c068156a`): `fetch_updated_dependency_files` no longer raises when the lockfile is unchanged. A constraint-only update (bumping a lower bound without affecting resolution) is a valid change — we return just the updated pyproject.toml without the lockfile. Added a test covering this scenario.

**Review by robaiken — APPROVED** [review]
